_posts/2020-02-19-angular-app-initializer.md

```diff
@@ -0,0 +1,174 @@
+  }]
+// (...)
+})
+export class AppModule { }
+```
+Powyższy kod wyświetli na konsoli konfigurację z pliku conf.json.
+```
+{name: "Test App name"}
+```
+
+## Implementacja w Angularze
+Przyjrzyjmy się, jak `APP_INITIALIZER` został zaimplementowany w samym Angularze.
+W pliku `application_init.ts` znajduje się definicja `InjectionToken`.
+```js
+export const APP_INITIALIZER = new InjectionToken<Array<() => void>>('Application Initializer');
+```
+Początek bootstrapowania aplikacji w Angular, wygląda następująco:
+##### *`platformRef#bootstrapModuleFactory()`*
+``` js
+return _callAndReportToErrorHandler(exceptionHandler, ngZone !, () => {
+       const initStatus: ApplicationInitStatus = moduleRef.injector.get(ApplicationInitStatus);
+       initStatus.runInitializers(); // (1)
+       return initStatus.donePromise.then(() => {
+         this._moduleDoBootstrap(moduleRef); // (2)
+         return moduleRef;
+       });
+```
+W punkcie (1) w serwisie ApplicationInitStatus wywołana jest funkcja runInitializers. Po zakończeniu ApplicationInitStatus, Angular przeprowadza bootstrap komponentu.
+Metoda runInitializers, sprawdza które wywołania zwróciły Promise i czeka aż wszystkie funkcje zostaną zakończone (resolve).
+##### *`ApplicationInitStatus#runInitializers()`*
+```js
+  runInitializers() {
+    // (...)
+    const asyncInitPromises: Promise<any>[] = [];
+
+    if (this.appInits) {
+      for (let i = 0; i < this.appInits.length; i++) {
+        const initResult = this.appInits[i]();
+        if (isPromise(initResult)) {
+          asyncInitPromises.push(initResult);
+        }
+      }
+    }
+
+    Promise.all(asyncInitPromises).then(() => { complete(); }).catch(e => { this.reject(e); });
+
+    // (...)
+  }
+```
+
+## Zastosowania
+Do czego można zastosować `APP_INITIALIZER`?
+* Obsługa powiadomień push z serwera (comety)
+* Pobranie konfiguracji np. CSRF token
+* Monitorowanie aktywności użytkownika
+* Keep alive
+* Keycloak - polecam świety wpis Michała na ten temat [źródło](https://blog.consdata.tech/2020/02/05/keycloak-z-angularem-i-keycloakiem.html)
+
+Nawet, jeżeli w swojej aplikacji nie używamy `APP_INITIALIZER`, sam Angular wykorzystuje go do poprawnego działania.
+Pryzkłady użycia w Angularze:
+* Routing (RouterModule), używany jest do poprawnej pracy Guard'ów.  [źródło](https://github.com/angular/angular/blob/e35d9eaa7d5267e9ea4d3fe2b85b88e28aae3f22/packages/router/src/router_module.ts#L510)
+* Web Worker (WorkerAppModule) [źródło](https://github.com/angular/angular/blob/8.2.x/packages/docs/web_workers/web_workers.md)
+* ServiceWorkerModule [źródło](https://github.com/angular/angular/blob/8b88269ae1c0d609e098964e60d08e8472f5aa40/packages/service-worker/src/module.ts#L161)
+* NgProbe. Angular 9 wprowadza [Ivy](https://angular.io/guide/ivy) i mechanizm NgProbe przestanie działać. [źródło](https://github.com/angular/angular/blob/8b88269ae1c0d609e098964e60d08e8472f5aa40/packages/platform-browser/src/dom/debug/ng_probe.ts#L41)
```

Comment:
W przypadku wyliczania, trzeba każdy przykład zakończyć przecinkiem, a w Twoim przypadku średnikiem, aby nie dublować przecinków. A na samym końcu kropka.
https://www.ekorekta24.pl/interpunkcja-w-wyliczeniach-przecinki-sredniki-a-moze-nic/

Reply:
omg zaraz musze to przetworzyc :o

Reply:
ok zrobilem